### PR TITLE
Offchain reading: merkle proof validation

### PR DESCRIPTION
**Pull request by pikonha**
# L2 offchain storage

## All Submissions:

* [X] Have you followed the guidelines in our Contributing document?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

## Description

All the required contracts for both L1 and L2 have been added to the repository.

## Related Issue

#66 

## Changes
- [X] New feature implementation
- [ ] Bug fix
- [ ] Code refactoring
- [ ] Documentation update
- [ ] Other (please specify)

### Changes to Core Features:

* [X] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [X] Have you successfully run tests with your changes locally?

## Additional Notes

No deployment script was added for now, only local deployment is supported.

### CODE REVIEW

**Review by LeonardoVieira1630 — COMMENTED** [review]
Hey lucas, i had some suggestions, please take a look before the merge = ) .

File: packages/contracts/foundry.toml [hunk 1]
```diff
@@ -2,11 +2,13 @@
 src = "src"
 out = "out"
 libs = ["lib"]
+viaIR = true
```
[the remark this answers is not in the document]
Comment (pikonha) [reply packages/contracts/foundry.toml:5]: it is an optimization setting the enables calls with a huge number of arguments such as:

```solidity
function _addr(bytes32 node, uint256 coinType, address target) private view returns (bytes memory) {
        EVMFetcher.newFetchRequest(verifier, target).getStatic(RECORD_VERSIONS_SLOT).element(node).getDynamic(
            VERSIONABLE_ADDRESSES_SLOT
        ).ref(0).element(node).element(coinType).fetch(this.addrCoinTypeCallback.selector, "");
    }
```

File: packages/contracts/package.json [hunk 2]
```diff
@@ -6,8 +6,9 @@
     "test:watch": "forge test -vvv --watch",
     "start:l1": "anvil & copyfiles ../../.env contracts/.env && forge script script/PublicResolver.s.sol --rpc-url http://localhost:8545 --broadcast -vv",
     "start:db": "anvil & copyfiles ../../.env contracts/.env && forge script script/DatabaseResolver.s.sol --rpc-url http://localhost:8545 --broadcast -vv",
+    "start:arb": "anvil & copyfiles ../../.env contracts/.env && forge script script/OffchainResolver.s.sol --rpc-url http://localhost:8545 --broadcast -vv",
```
[the remark this answers is not in the document]
Comment (pikonha) [reply packages/contracts/package.json:9]: this is the script for running locally which has been changed on #69

File: packages/contracts/script/OffchainResolver.s.sol [hunk 3]
```diff
@@ -0,0 +1,48 @@
+// SPDX-License-Identifier: UNLICENSED
```
[the remark this answers is not in the document]
Comment (pikonha) [reply packages/contracts/script/OffchainResolver.s.sol:1]: yeah, this is the minimal setup for handling read and validation through Merkle Proofs.

File: packages/contracts/src/ArbitrumResolver.sol [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    DNSResolver,
+    InterfaceResolver,
+    NameResolver,
+    PubkeyResolver,
+    TextResolver
+{
+    mapping(bytes32 => address) private _owners;
+
+    function isAuthorised(bytes32 node) internal view override returns (bool) {
+        // console.log("_owners[node]", _owners[node]);
```
[the remark this answers is not in the document]
Comment (pikonha) [reply packages/contracts/src/ArbitrumResolver.sol]: fixed and renamed to L2Resolver

File: packages/contracts/src/evmgateway/L1Resolver.sol [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    uint256 constant COIN_TYPE_ETH = 60;
+    uint256 constant RECORD_VERSIONS_SLOT = 0;
+    uint256 constant VERSIONABLE_ADDRESSES_SLOT = 2;
+    uint256 constant VERSIONABLE_HASHES_SLOT = 3;
+    uint256 constant VERSIONABLE_TEXTS_SLOT = 10;
+
+    event TargetSet(bytes32 indexed node, address target);
+
+    function isAuthorised(bytes32 node) internal view returns (bool) {
+        // TODO: Add support for
```
[the remark this answers is not in the document]
Comment (pikonha) [reply packages/contracts/src/evmgateway/L1Resolver.sol:34]: it is still pending to support

**Review by LeonardoVieira1630 — APPROVED** [review]

**Pull request merged**